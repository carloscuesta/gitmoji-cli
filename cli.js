#!/usr/bin/env node
'use strict';

const meow = require('meow');
const axios = require('axios');
const updateNotifier = require('update-notifier');
const GitmojiCli = require('./src/gitmoji.js');
const pkg = require('./package.json');

updateNotifier({pkg}).notify();

const cli = meow(`
	Usage
	  $ gitmoji
	Options
		--init, -i	Initialize gitmoji as a commit hook
		--commit, -c Interactively commit using the prompts
		--list, -l  List all the available gitmojis
		--search, -s	Search gitmojis
		--version, -v	Print gitmoji-cli installed version
		--update, -u	Sync emoji list with the repo
	Examples
		$ gitmoji -l
		$ gitmoji bug linter -s
`, {
	alias: {
		i: 'init',
		c: 'commit',
		l: 'list',
		s: 'search',
		h: 'help',
		v: 'version',
		u: 'update'
	}
});

const gitmojiApiClient = axios.create({
	baseURL: 'https://raw.githubusercontent.com/carloscuesta/gitmoji/master',
	timeout: 5000,
	headers: {},
	params: {}
});

const gitmojiCli = new GitmojiCli(gitmojiApiClient);

const commands = {
	list: () => gitmojiCli.list(),
	search: () => cli.input.map(element => gitmojiCli.search(element)),
	init: () => gitmojiCli.init(),
	hook: () => gitmojiCli.ask('hook'),
	version: () => console.log(gitmojiCli.version(pkg.version)),
	commit: () => gitmojiCli.ask('client'),
	update: () => gitmojiCli.updateCache(),
	undefined: () => {
		if (process.argv[2] === '--hook') {
			gitmojiCli.ask('hook');
		} else {
			gitmojiCli.ask('client');
		}
	}
};

const arg = Object.keys(cli.flags)[1];
if (commands[arg]) {
	commands[arg]();
} else {
	cli.showHelp();
}
