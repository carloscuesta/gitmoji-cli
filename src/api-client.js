const axios = require('axios');

module.exports = gitmojiApiClient;

function gitmojiApiClient(baseURL) {
	return axios.create({
		baseURL,
		timeout: 5000,
		headers: {},
		params: {}
	});
}
