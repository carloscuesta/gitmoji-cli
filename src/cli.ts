#!/usr/bin/env node
import meow from 'meow'
import updateNotifier from 'update-notifier'
import { readFileSync } from 'fs'

import FLAGS from '@constants/flags'
import findGitmojiCommand from '@utils/findGitmojiCommand'
import type { CommitOptions } from '@commands/commit'
import type { SearchOptions } from '@commands/search'

const packageJson = readFileSync(
  new URL('../package.json', import.meta.url)
).toString()

updateNotifier({ pkg: JSON.parse(packageJson) }).notify({
  isGlobal: true
})

const cli = meow(
  `
  Usage
    $ gitmoji [option] [command]
  Options
    --${FLAGS.COMMIT}, -c    Interactively commit using the prompts
    --${FLAGS.CONFIG}, -g    Setup gitmoji-cli preferences.
    --${FLAGS.INIT}, -i      Initialize gitmoji as a commit hook
    --${FLAGS.LIST}, -l      List all the available gitmojis
    --${FLAGS.REMOVE}, -r    Remove a previously initialized commit hook
    --${FLAGS.SEARCH}, -s    Search gitmojis
    --${FLAGS.UPDATE}, -u    Sync emoji list with the repo
    --${FLAGS.VERSION}, -v   Print gitmoji-cli installed version
  Commands
    commit          Interactively commit using the prompts
    config          Setup gitmoji-cli preferences.
    init            Initialize gitmoji as a commit hook
    list            List all the available gitmojis
    remove          Remove a previously initialized commit hook
    search          Search gitmojis
    update          Sync emoji list with the repo
  Examples
    $ gitmoji -l
    $ gitmoji bug linter -s
`,
  {
    importMeta: { url: import.meta.url } as ImportMeta,
    flags: {
      [FLAGS.COMMIT]: { type: 'boolean', alias: 'c' },
      [FLAGS.CONFIG]: { type: 'boolean', alias: 'g' },
      [FLAGS.HELP]: { type: 'boolean', alias: 'h' },
      [FLAGS.INIT]: { type: 'boolean', alias: 'i' },
      [FLAGS.LIST]: { type: 'boolean', alias: 'l' },
      [FLAGS.REMOVE]: { type: 'boolean', alias: 'r' },
      [FLAGS.SEARCH]: { type: 'boolean', alias: 's' },
      [FLAGS.UPDATE]: { type: 'boolean', alias: 'u' },
      [FLAGS.VERSION]: { type: 'boolean', alias: 'v' }
    }
  }
)

export const options = {
  [FLAGS.COMMIT]: async (options: CommitOptions) =>
    await (await import('@commands/commit')).default(options),
  [FLAGS.CONFIG]: async () =>
    await (await import('@commands/config')).default(),
  [FLAGS.HOOK]: async (options: CommitOptions) =>
    await (await import('@commands/commit')).default(options),
  [FLAGS.INIT]: async () =>
    await (await import('@commands/hook')).default.create(),
  [FLAGS.LIST]: async () => await (await import('@commands/list')).default(),
  [FLAGS.REMOVE]: async () =>
    await (await import('@commands/hook')).default.remove(),
  [FLAGS.SEARCH]: async (options: SearchOptions) =>
    await (await import('@commands/search')).default(options),
  [FLAGS.UPDATE]: async () => await (await import('@commands/update')).default()
}

findGitmojiCommand(cli, options)
