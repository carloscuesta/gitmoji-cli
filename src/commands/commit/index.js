// @flow
import inquirer from 'inquirer'

import getEmojis from '../../utils/getEmojis'
import prompts from './prompts'
import withHook from './withHook'
import withClient from './withClient'

const commit = (mode: 'client' | 'hook') => {
  return getEmojis()
    .then((gitmojis) => prompts(gitmojis))
    .then((questions) => {
      inquirer.prompt(questions).then((answers) => {
        if (mode === 'hook') return withHook(answers)

        return withClient(answers)
      })
    })
}

export default commit
