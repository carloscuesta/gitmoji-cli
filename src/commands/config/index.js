// @flow
import inquirer from 'inquirer'

import configurationPrompts, { CONFIGURATION_PROMPT_NAMES } from './prompts'
import configurationVault from '../../utils/configurationVault'

const config = () => {
  inquirer.prompt(configurationPrompts()).then((answers) => {
    configurationVault.setAutoAdd(answers[CONFIGURATION_PROMPT_NAMES.AUTO_ADD])
    configurationVault.setEmojiFormat(
      answers[CONFIGURATION_PROMPT_NAMES.EMOJI_FORMAT]
    )
    configurationVault.setSignedCommit(
      answers[CONFIGURATION_PROMPT_NAMES.SIGNED_COMMIT]
    )
    configurationVault.setScopePrompt(
      answers[CONFIGURATION_PROMPT_NAMES.SCOPE_PROMPT]
    )
  })
}

export default config
