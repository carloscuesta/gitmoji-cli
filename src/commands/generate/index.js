// @flow
import inquirer from 'inquirer'
import getEmojis from '@utils/getEmojis'
import getGeneratedMessage from '@utils/getGeneratedMessage'
import { execa } from 'execa'
import { type ParsedMessage } from '@utils/getGeneratedMessage'
import chalk from 'chalk'
import withClient from '@commands/commit/withClient'

type CommitOptions = {
  model?: 'gpt-3.5-turbo' | 'gpt-4',
  context?: boolean
}

const generate = async (options: CommitOptions): Promise<void> => {
  if (!process.env.OPENAI_API_KEY) {
    return console.error(
      chalk.red(`Unable to locate OPENAI_API_KEY in environment`)
    )
  }

  const { stdout: diff } = await execa('git', ['diff', '--cached'])
  if (!diff) {
    return console.error(chalk.red('Could not find any staged files to commit'))
  }

  // Let the user specify additional context to give some guidance to the generated message
  let context
  if (!!options.context) {
    ;({ context } = await inquirer.prompt(
      {
        type: 'text',
        name: 'context',
        message: 'Provide additional context for the commit message'
      },
      {
        onCancel: () => process.exit()
      }
    ))
  }

  let message: ParsedMessage | void
  const gitmojis = await getEmojis()
  while (true) {
    message = await getGeneratedMessage(diff, gitmojis, context, options.model)
    if (!message) {
      return
    }

    const { confirm } = await inquirer.prompt({
      type: 'confirm',
      name: 'confirm',
      message: message.raw
    })

    if (confirm) {
      break
    }
  }

  // Sanity should not be able to get here but Flow needed it
  if (!message) {
    return console.error(chalk.red('Unable to generate commit message'))
  }

  await withClient({
    title: message.title,
    gitmoji: message.gitmoji,
    message: ''
  })
}

export default generate
