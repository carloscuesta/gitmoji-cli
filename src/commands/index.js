const search = require('./search').default
const list = require('./list').default

export default {
  search,
  list
}
