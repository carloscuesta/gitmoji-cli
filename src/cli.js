#!/usr/bin/env node
import meow from 'meow'
import updateNotifier from 'update-notifier'

import pkg from '../package.json'
import commands from './commands'
import findGitmojiCommand from './utils/findGitmojiCommand'

updateNotifier({ pkg }).notify()

const cli = meow(
  `
  Usage
    $ gitmoji
  Options
    --commit, -c    Interactively commit using the prompts
    --config, -g    Setup gitmoji-cli preferences.
    --init, -i      Initialize gitmoji as a commit hook
    --list, -l      List all the available gitmojis
    --remove, -r    Remove a previously initialized commit hook
    --search, -s    Search gitmojis
    --update, -u    Sync emoji list with the repo
    --version, -v   Print gitmoji-cli installed version
  Examples
    $ gitmoji -l
    $ gitmoji bug linter -s
`,
  {
    flags: {
      commit: { type: 'boolean', alias: 'c' },
      config: { type: 'boolean', alias: 'g' },
      help: { type: 'boolean', alias: 'h' },
      init: { type: 'boolean', alias: 'i' },
      list: { type: 'boolean', alias: 'l' },
      remove: { type: 'boolean', alias: 'r' },
      search: { type: 'boolean', alias: 's' },
      update: { type: 'boolean', alias: 'u' },
      version: { type: 'boolean', alias: 'v' }
    }
  }
)

export const options = {
  commit: () => commands.commit('client'),
  config: () => commands.config(),
  hook: () => commands.commit('hook'),
  init: () => commands.createHook(),
  list: () => commands.list(),
  remove: () => commands.removeHook(),
  search: () => cli.input.map((input) => commands.search(input)),
  update: () => commands.update()
}

findGitmojiCommand(cli, options)
