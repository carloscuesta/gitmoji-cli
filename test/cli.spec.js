import updateNotifier from 'update-notifier'
import meow from 'meow'

import pkg from '../package.json'
import { options } from '../src/cli'
import commands from '../src/commands'

jest.mock('../src/commands')

describe('cli', () => {
  it('should call updateNotifier', () => {
    expect(updateNotifier).toHaveBeenCalledWith({ pkg })
    expect(updateNotifier({ pkg }).notify).toHaveBeenCalled()
  })

  it('should match meow with cli information', () => {
    expect(meow.mock.calls).toMatchSnapshot()
  })

  it('should call commit command on commit', () => {
    options.commit('client')
    expect(commands.commit).toHaveBeenCalledWith('client')
  })

  it('should call config command on config', () => {
    options.config()
    expect(commands.config).toHaveBeenCalledWith()
  })

  it('should call commit command on hook', () => {
    options.hook('hook')
    expect(commands.commit).toHaveBeenLastCalledWith('hook')
  })

  it('should call createHook command on init', () => {
    options.init()
    expect(commands.createHook).toHaveBeenCalledWith()
  })

  it('should call list command on list', () => {
    options.list()
    expect(commands.list).toHaveBeenCalledWith()
  })

  it('should call removeHook command on remove', () => {
    options.remove()
    expect(commands.removeHook).toHaveBeenCalledWith()
  })

  it('should call search command on search', () => {
    options.search()
    expect(commands.search).toHaveBeenCalledWith('testSearchQuery')
  })

  it('should call update command on update', () => {
    options.update()
    expect(commands.search).toHaveBeenCalledWith('testSearchQuery')
  })
})
