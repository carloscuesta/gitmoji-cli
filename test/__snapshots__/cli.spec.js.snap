// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cli should match meow with cli information 1`] = `
Array [
  Array [
    "
  Usage
    $ gitmoji
  Options
    --commit, -c    Interactively commit using the prompts
    --config, -g    Setup gitmoji-cli preferences.
    --init, -i      Initialize gitmoji as a commit hook
    --list, -l      List all the available gitmojis
    --remove, -r    Remove a previously initialized commit hook
    --search, -s    Search gitmojis
    --update, -u    Sync emoji list with the repo
    --version, -v   Print gitmoji-cli installed version
  Examples
    $ gitmoji -l
    $ gitmoji bug linter -s
",
    Object {
      "flags": Object {
        "commit": Object {
          "alias": "c",
          "type": "boolean",
        },
        "config": Object {
          "alias": "g",
          "type": "boolean",
        },
        "help": Object {
          "alias": "h",
          "type": "boolean",
        },
        "init": Object {
          "alias": "i",
          "type": "boolean",
        },
        "list": Object {
          "alias": "l",
          "type": "boolean",
        },
        "remove": Object {
          "alias": "r",
          "type": "boolean",
        },
        "search": Object {
          "alias": "s",
          "type": "boolean",
        },
        "update": Object {
          "alias": "u",
          "type": "boolean",
        },
        "version": Object {
          "alias": "v",
          "type": "boolean",
        },
      },
    },
  ],
]
`;
